File: src/main/java/net/ndrei/teslacorelib/inventory/SidedItemHandler.java
package net.ndrei.teslacorelib.inventory;

import net.minecraft.item.EnumDyeColor;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.items.IItemHandler;
import net.ndrei.teslacorelib.capabilities.inventory.SidedItemHandlerConfig;

import java.util.Arrays;
import java.util.List;

/**
 * Created by CF on 2016-12-15.
 */
public class SidedItemHandler extends MultiItemHandler implements ISidedItemHandler {
    private SidedItemHandlerConfig sidedConfig = null;

    public SidedItemHandler(SidedItemHandlerConfig config) {
        this(null, config);
    }

    public SidedItemHandler(List<IItemHandler> handlers, SidedItemHandlerConfig config) {
        super(handlers);

        // this.facesConfig = new HashMap<>();
        this.sidedConfig = config;
    }

    @Override
    public int[] getSlotsForFace(EnumFacing side) {
        int[] result = new int[0];

        int index = 0;
        for(int i = 0; i < this.getInventories(); i++) {
            IItemHandler handler = this.getInventory(i);
            int size = handler.getSlots();
            if (handler instanceof ColoredItemHandler) {
                EnumDyeColor color = ((ColoredItemHandler)handler).getColor();
                if (this.sidedConfig.isSideSet(color, side)) {
                    int startIndex = result.length;
                    result = Arrays.copyOf(result, result.length + size);
                    for(int x = 0; x < size; x++) {
                        result[startIndex + x] = index + x;
                    }
                }
            }
            index += size;
        }

        return result;
    }

    public SidedItemHandlerWrapper getSideWrapper(EnumFacing side) {
        // TODO: cache them!
        return new SidedItemHandlerWrapper(this, side);
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/inventory/SidedItemHandler.kt
package net.ndrei.teslacorelib.inventory

import net.minecraft.util.EnumFacing
import net.minecraftforge.items.IItemHandler
import net.ndrei.teslacorelib.capabilities.inventory.SidedItemHandlerConfig
import java.util.*

/**
 * Created by CF on 2017-06-28.
 */
class SidedItemHandler(handlers: MutableList<IItemHandler>?, private val sidedConfig: SidedItemHandlerConfig)
    : MultiItemHandler(handlers ?: mutableListOf()), ISidedItemHandler {
    constructor(sidedConfig: SidedItemHandlerConfig) : this(null, sidedConfig) {}

    override fun getSlotsForFace(side: EnumFacing): IntArray {
        var result = IntArray(0)

        var index = 0
        for (i in 0..this.inventories - 1) {
            val handler = this.getInventory(i)
            val size = handler.slots
            if (handler is ColoredItemHandler) {
                val color = handler.color
                if (this.sidedConfig!!.isSideSet(color, side)) {
                    val startIndex = result.size
                    result = Arrays.copyOf(result, result.size + size)
                    for (x in 0..size - 1) {
                        result[startIndex + x] = index + x
                    }
                }
            }
            index += size
        }

        return result
    }

    fun getSideWrapper(side: EnumFacing): SidedItemHandlerWrapper {
        // TODO: cache them!
        return SidedItemHandlerWrapper(this, side)
    }
}
